File: src/main/java/no/nav/pensjondokdist/saf/SafClient.java
package no.nav.pensjondokdist.saf;

import static no.nav.pensjondokdist.util.JsonUtil.toJsonString;
import static no.nav.pensjondokdist.util.JsonUtil.toObjectFromJsonString;

import java.io.IOException;
import java.util.List;
import java.util.StringJoiner;
import java.util.stream.Collectors;

import com.fasterxml.jackson.core.type.TypeReference;

import no.nav.pensjondokdist.saf.model.Journalpost;
import org.apache.commons.codec.Charsets;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.util.StreamUtils;
import org.springframework.web.client.RestTemplate;

import no.nav.pensjondokdist.PensjonDokdistException;
import no.nav.pensjondokdist.oidc.OidcProvider;
import no.nav.pensjondokdist.saf.graphql.GraphQlError;
import no.nav.pensjondokdist.saf.graphql.GraphQlRequest;
import no.nav.pensjondokdist.saf.graphql.GraphQlResponse;
import no.nav.pensjondokdist.saf.graphql.Variables;

@Component
public class SafClient {
    private String url;
    private String query;
    private RestTemplate restTemplate;
    private OidcProvider oidcProvider;
    private static Logger logger = LoggerFactory.getLogger(SafClient.class);

    public SafClient(@Value("${saf.graphql}") String url, RestTemplate restTemplate, OidcProvider oidcProvider) throws IOException {
        this.url = url;
        this.restTemplate = restTemplate;
        this.oidcProvider = oidcProvider;
        query = StreamUtils.copyToString(new ClassPathResource("saf/journalpostQuery.graphql").getInputStream(), Charsets.UTF_8);
    }

    Journalpost fetchJournal(String journalpostId) {
        GraphQlRequest graphQlRequest = new GraphQlRequest(query, new Variables(journalpostId));
        try {
            ResponseEntity<String> entity = restTemplate.exchange(url, HttpMethod.POST,
                    new HttpEntity<>(toJsonString(graphQlRequest), oidcProvider.getHeaders()),
                    String.class);
            return handleResponse(entity, journalpostId);
        } catch (Exception e) {
            logger.error("Kunne ikke hente journalpost for journalpostId " + journalpostId);
            throw new PensjonDokdistException("Kunne ikke hente journalpost for journalpostId " + journalpostId, e);
        }
    }

    private Journalpost handleResponse(ResponseEntity<String> entity, String journalpostId) {
        if (entity.getStatusCode() == HttpStatus.OK && entity.getBody() != null) {
            GraphQlResponse graphQlResponse = toObjectFromJsonString(entity.getBody(), new TypeReference<GraphQlResponse>() {
            });
            if (graphQlResponse.getData() == null || graphQlResponse.getData().getJournalpost() == null) {
                List<String> errorMessageList = graphQlResponse.getErrors()
                        .stream()
                        .map(GraphQlError::getMessage)
                        .collect(Collectors.toList());
                StringJoiner stringJoiner = new StringJoiner(System.lineSeparator());
                stringJoiner.add("Journalpost for journalpostId " + journalpostId + " er null.");
                errorMessageList.forEach(stringJoiner::add);
                String errors = errorMessageList.toString();
                logger.error(errors);
                throw new PensjonDokdistException(errors);
            }
            return graphQlResponse.getData().getJournalpost();
        } else {
            logger.error("Fant ikke journalpost: " + journalpostId);
            throw new PensjonDokdistException("Fant ikke journalpost. StatusCode: " + entity.getStatusCode().value());
        }
    }
}

File: src/main/kotlin/no/nav/pensjon/dokdist/saf/SafClient.kt
package no.nav.pensjon.dokdist.saf

import com.fasterxml.jackson.databind.ObjectMapper
import net.logstash.logback.marker.RawJsonAppendingMarker
import no.nav.pensjon.dokdist.auth.*
import no.nav.pensjon.dokdist.graphql.*
import org.slf4j.LoggerFactory
import org.springframework.boot.restclient.RestTemplateBuilder
import org.springframework.core.io.ClassPathResource
import org.springframework.http.*
import org.springframework.util.StreamUtils
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.client.HttpClientErrorException.BadRequest
import org.springframework.web.client.HttpClientErrorException.Forbidden
import org.springframework.web.client.HttpClientErrorException.NotFound
import org.springframework.web.client.RestClientException
import org.springframework.web.client.exchange
import org.springframework.web.server.ResponseStatusException
import java.nio.charset.StandardCharsets

interface SafService {
    fun fetchJournal(journalpostId: String): Journalpost?
}

@ResponseStatus(code = HttpStatus.INTERNAL_SERVER_ERROR, reason = "Feilet å hente journalpost i SAF")
class SafException(message: String?, cause: Throwable? = null) : RuntimeException(message, cause)

private data class FetchJournalVariables(val journalpostId: String)
private data class FetchJournalData(val journalpost: Journalpost?)


class SafClient(
    private val url: String,
    accessToken: OnBehalfOfTokenResponse,
    restTemplateBuilder: RestTemplateBuilder = RestTemplateBuilder(),
) : AzureAdOBOApiBinding(accessToken, restTemplateBuilder), SafService {
    private val query: String = ClassPathResource("saf/journalpostQuery.graphql").inputStream.use { StreamUtils.copyToString(it, StandardCharsets.UTF_8) }

    override fun fetchJournal(journalpostId: String): Journalpost? {
        return try {
            val response = restTemplate.exchange<GraphQLResponse<FetchJournalData>>(
                "$url/graphql",
                HttpMethod.POST,
                HttpEntity(
                    GraphQLRequest(query, FetchJournalVariables(journalpostId)),
                    HttpHeaders().apply {
                        contentType = MediaType.APPLICATION_JSON
                        add(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE)
                    }
                ),
            ).body
            handleResponse(response, journalpostId)
        } catch (e: NotFound) {
            null
        } catch (e: Forbidden) {
            val msg = "Mangler tilgang til journalpost: $journalpostId"
            logger.info(
                RawJsonAppendingMarker("error_response", e.responseBodyAsString),
                msg,
                e
            )
            throw ResponseStatusException(HttpStatus.FORBIDDEN, msg)
        } catch (e: BadRequest) {
            val msg = "Feil i graphql-spørring til saf for journalpost: $journalpostId"
            logger.error(
                RawJsonAppendingMarker("error_response", e.responseBodyAsString),
                msg,
                e
            )
            throw SafException(msg, e)
        } catch (e: RestClientException) {
            val msg = "Kunne ikke hente journalpost for journalpostId $journalpostId: ukjent grunn"
            logger.error(msg, e)
            throw SafException(msg, e)
        }
    }

    private fun handleResponse(response: GraphQLResponse<FetchJournalData>?, journalpostId: String): Journalpost? {
        return if (response == null) {
            val msg = "Fikk tomt svar fra saf.journalpost for: $journalpostId"
            logger.error(msg)
            throw SafException(msg)
        } else if (response.data == null) {
            logger.error(
                RawJsonAppendingMarker(
                    "error_response", objectMapper.writeValueAsString(response.errors)
                ), "Journalpost GraphQL spørring feilet for: $journalpostId"
            )

            throw SafException("Hent journalpost feilet: $journalpostId")
        } else if (response.data.journalpost == null) {
            logger.warn("Journalpost mangler i svar fra SAF, men fikk ikke notFound som forventet.")
            null
        } else {
            response.data.journalpost
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(SafClient::class.java)
        private val objectMapper = ObjectMapper()
    }
}
